client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala

```diff
@@ -499,62 +564,84 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
             s" shuffleId: $shuffleId $latestLoc")
           return
         }
-        val set = new util.HashSet[RpcCallContext]()
-        set.add(context)
+        val set = new util.HashSet[ChangePartitionRequest]()
+        set.add(changePartition)
         requests.put(partitionId, set)
       }
     }
+  }
 
-    if (cause.isDefined) {
-      blacklistPartition(shuffleId, oldPartition, cause.get)
+  def batchHandleChangePartitions(
+      applicationId: String,
+      shuffleId: Int,
+      changePartitions: Array[ChangePartitionRequest]): Unit = {
+    val requests = changePartitionRequests.get(shuffleId)
+
+    // Blacklist all failed workers
+    if (changePartitions.exists(_.causes.isDefined)) {
+      changePartitions.filter(_.causes.isDefined).foreach { changePartition =>
+        blacklistPartition(shuffleId, changePartition.oldPartition, changePartition.causes.get)
+      }
+    }
+
+    // remove together to reduce lock time
+    def replySuccess(locations: Array[PartitionLocation]): Unit = {
+      requests.synchronized {
```

Comment:
Better to rename requests to requestsMap

Reply:
Done